### PR TITLE
IncludingFile: Add custom keywords to detect to reduce false positives

### LINKED ISSUE #407: Minimize false positives on WordPressVIPMinimum.Files.IncludingFile.IncludingFile
`require_once \PL\PATH . '/inc/endpoints/especiales-endpoint.php';`

`❗️ Warning: Absolute include path must be used. Use get_template_directory(), get_stylesheet_directory() or plugin_dir_path() (WordPressVIPMinimum.Files.IncludingFile.IncludingFile).`

## What problem would the enhancement address for VIP?
Currently clients include files be defining path variables or constants that get used for the path. This makes the IncludingFile rule very noisy with false postiives.

## Describe the solution you'd like
We should enhance the rule check to account for path variables/constants to reduce false positives.

### PR DESCRIPTION
**Pull request by rebeccahum**
Fixes #407.

Takes into account for URLs and $restrictedConstants with keywords from $customPaths.

### CODE REVIEW

File: WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php [hunk 1]
```diff
@@ -104,6 +115,16 @@ public function process_token( $stackPtr ) {
 			return;
 		}
 
+		if ( $this->tokens[ $nextToken ]['code'] === T_CONSTANT_ENCAPSED_STRING && filter_var( str_replace( [ '"', "'" ], '', $this->tokens[ $nextToken ]['content'] ), FILTER_VALIDATE_URL ) ) {
```
Comment (jrfnl) [inline WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php:118]: Why did you move this check up ?

File: WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php [hunk 2]
```diff
@@ -104,6 +115,16 @@ public function process_token( $stackPtr ) {
...
+			$message = 'Include path must be local file source, external URLs are prohibited on WordPress VIP.';
+			$this->phpcsFile->addError( $message, $nextToken, 'ExternalURL' );
+			return;
+		}
+
+		if ( $this->has_custom_path( $nextToken ) !== false && array_key_exists( $this->tokens[ $nextToken ]['content'], $this->restrictedConstants ) === false ) {
```
Comment (jrfnl) [inline WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php:124]: * Why are you checking against `!== false` when the method will only ever return `true` or `false` ?
* Why use `array_key_exists()` instead of `isset()` ?
* Why are these conditions in this order ?

File: WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php [hunk 3]
```diff
@@ -163,13 +184,35 @@ public function process_token( $stackPtr ) {
...
+	 *                         This parameter needs to be passed if no $content is
+	 *                         passed.
+	 * @param string $content  Optionally, the current content string, might be a
+	 *                         substring of the original string.
+	 *                         Defaults to `false` to distinguish between a passed
+	 *                         empty string and not passing the $content string.
+	 *
+	 * @return bool True if the string contains a keyword in $customPaths, false otherwise.
+	 */
+	public function has_custom_path( $stackPtr = null, $content = null ) {
```
Comment (jrfnl) [inline WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php:205]: * Why is this a `public` method ?
* Why are both parameters optional ?
* Why is no check being done against the token code/type ? This check will now also bail out on a call to `dirname()` or on the variable `$custom_path` which is not the intention of this check.

File: WordPressVIPMinimum/Tests/Files/IncludingFileUnitTest.inc [hunk 4]
```diff
@@ -13,6 +13,10 @@ require dirname( __FILE__, 2 ) . '/my_file.php' );
 require dirname( dirname( __FILE__ ) ) . '/my_file.php' );
 require dirname( dirname( __DIR__ ) ) . '/my_file.php' );
 require_once( get_parent_theme_file_path( '/my_file.php' ); );
+require_once(    $custom_path );
```
Comment (jrfnl) [inline WordPressVIPMinimum/Tests/Files/IncludingFileUnitTest.inc:16]: This one should trigger the `UsingVariable` error code, but doesn't.

**Review by jrfnl — COMMENTED** [review]
Hi Rebecca,

I've been looking at this change and notice I find it difficult to review it as there is so much more which needs improving in this sniff, that it is hard to focus on this specific change.

This might actually be a good sniff to take the time to dissect in detail in a live coding call and talk through every code inefficiency and assumption in the code which should be fixed.

For now, specifically for this PR, let's start with some questions (left in-line)...

File: WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php (hunk 1 above)
Comment (rebeccahum) [reply WordPressVIPMinimum/Sniffs/Files/IncludingFileSniff.php:118]: I moved it up so that it doesn't bail before checking for an external URL with a keyword from `$customPaths` in it. e.g.:

```
include( "http://path.com/bad_file.php" );
```

**Comment by rebeccahum** [comment]
Closing for now as per our discussion on making this an atomic commit for the next release and then doing more re-factoring for 3.0.
